pkg/controller/addon/addon_controller.go

```diff
@@ -0,0 +1,358 @@
+		})
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	log.Info("successfully applied all resources")
+
+	if err := r.refreshCR(res); err != nil {
+		log.Error(err, "status update failed to refresh object")
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	err = r.updateStatus(res, op.AddonCondition{
+		Code: op.InstalledStatus, Version: res.Spec.Version})
+
+	//requeue true as isUptodate will be validated in the next reconcile loop
+	return reconcile.Result{Requeue: true}, err
+}
+
+func (r *ReconcileAddon) processPayload(res *op.Addon, targetNS string) (*mf.Manifest, error) {
+	addonPath := getAddonPath(res)
+	manifest, err := mf.NewManifest(addonPath, true, r.client)
+	if err != nil {
+		return nil, err
+	}
+
+	// set the currnet addon CRD instance as owner for all items in manifest
+	// set targetNamespace of the tekton-pipeline installation as the
+	// target namespace for the addon components
+	tfs := []mf.Transformer{
+		mf.InjectOwner(res),
+		mf.InjectNamespace(targetNS),
+	}
+
+	if err := manifest.Transform(tfs...); err != nil {
+		return nil, err
+	}
+	return &manifest, nil
+}
+
+func getAddonPath(res *op.Addon) string {
+	addonDir := getAddonBase(res)
+	path := filepath.Join(addonDir, res.Spec.Version)
+	return path
+}
+
+func GetLatestVersion(res *op.Addon) (string, error) {
+	dirName := getAddonBase(res)
+	items, err := ioutil.ReadDir(dirName)
+	if err != nil || len(items) == 0 {
+		return "", errAddonVersionUnresolved
+	}
+
+	return items[len(items)-1].Name(), nil
+}
+
+func getAddonBase(res *op.Addon) string {
+	return filepath.Join("deploy", "resources", "addons", res.Name)
+}
+
+func requestLogger(req reconcile.Request, context string) logr.Logger {
+	return ctrlLog.WithName(context).WithValues(
+		"Request.Namespace", req.Namespace,
+		"Request.NamespaceName", req.NamespacedName,
+		"Request.Name", req.Name)
+}
+
+// updateStatus set the status of res to s and refreshes res to the lastest version
+func (r *ReconcileAddon) updateStatus(res *op.Addon, c op.AddonCondition) error {
+
+	// NOTE: need to use a deepcopy since Status().Update() seems to reset the
+	// APIVersion of the res to "" making the object invalid; may be a mechanism
+	// to prevent us from using stale version of the object
+
+	tmp := res.DeepCopy()
+	tmp.Status.Conditions = append([]op.AddonCondition{c}, tmp.Status.Conditions...)
+
+	if err := r.client.Status().Update(context.TODO(), tmp); err != nil {
+		log.Error(err, "status update failed")
+		return err
+	}
+
+	if err := r.refreshCR(res); err != nil {
+		log.Error(err, "status update failed to refresh object")
+		return err
+	}
+	return nil
+}
+
+func (r *ReconcileAddon) refreshCR(res *op.Addon) error {
+	objKey := types.NamespacedName{
+		Namespace: res.Namespace,
+		Name:      res.Name,
+	}
+	return r.client.Get(context.TODO(), objKey, res)
+}
+
+func (r *ReconcileAddon) getPipelineRes() (*op.Config, error) {
+	res := &op.Config{}
+	namespacedName := types.NamespacedName{
+		Namespace: "",
```

Comment:
this can be removed - right? Default is ""

Reply:
Yes.